### PR TITLE
Add BatchInferenceTransform for efficient batch processing

### PR DESCRIPTION
**Pull request by marklubin**
## Summary

This PR introduces the `BatchInferenceTransform` component, a generic batch processing transform that optimizes LLM API usage through intelligent batching and retry logic.

### 🏗️ Architecture
- **Generic types** (`T`, `R`) for flexible input/output types
- **Protocol-based interfaces** for provider and storage backends
- **Configuration-driven** batch size and retry behavior

### 🔧 Core Components

1. **BatchInferenceTransform**
   - Partitions inputs into configurable batch sizes
   - Processes batches with automatic retry on failure
   - Tracks execution metrics and metadata
   - Returns results in original order

2. **Protocol Interfaces**
   - `BatchProvider`: Interface for batch-capable providers
   - `BatchStore`: Interface for result storage (caching ready)
   - `BatchConfig`: Configuration with batch size, retries, and hash function

3. **Error Handling**
   - Retry logic with exponential backoff
   - Configurable max retries and delays
   - Detailed error reporting in ComponentResult

### ✅ Testing
- 13 comprehensive unit tests covering:
  - Batch partitioning edge cases
  - Retry and failure scenarios
  - Large dataset processing
  - Custom processing functions
  - Execution time tracking

## Implementation Status

**Completed:**
- ✅ Core batch processing logic
- ✅ Retry mechanism with exponential backoff
- ✅ Error handling and reporting
- ✅ Comprehensive test coverage
- ✅ Integration with component pipeline system

**Planned (Next Iteration):**
- 🔄 Hash-based deduplication
- 🔄 Result caching with BatchStore
- 🔄 Skip processing for cached results

## Usage Example

```python
# Configure batch processing
config = BatchConfig(
    batch_size=10,
    hash_function=lambda x: hashlib.md5(x.encode()).hexdigest(),
    max_retries=3,
    retry_delay=1.0
)

# Create transform
transform = BatchInferenceTransform(
    provider=my_batch_provider,
    store=my_cache_store,
    config=config
)

# Process items in optimal batches
result = transform.process(items)
```

## Test Results
All tests passing:
- ✅ 13/13 unit tests
- ✅ Ruff linting clean
- ✅ Type annotations complete

🤖 Generated with [Claude Code](https://claude.ai/code)

### CODE REVIEW

File: apiana/core/components/transform/batch_inference.py [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+        """Retrieve cached result by hash key."""
+        ...
+    
+    def store_result(self, hash_key: str, input_item: T, result: R) -> None:
+        """Store computation result with hash key."""
+        ...
+    
+    def batch_get_by_hashes(self, hash_keys: List[str]) -> Dict[str, Optional[R]]:
+        """Batch retrieve multiple results by hash keys."""
+        ...
```
Comment (marklubin) [inline apiana/core/components/transform/batch_inference.py]: Let's remove and assume we dont have this and have to make individual calls.

File: apiana/core/components/transform/batch_inference.py [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+        self.provider = provider
+        self.store = store
+        self.batch_config = config
+        # Convert BatchConfig to dict for base class
+        config_dict = {
+            "batch_size": config.batch_size,
+            "max_retries": config.max_retries,
+            "retry_delay": config.retry_delay,
+            "timeout": config.timeout
+        }
```
Comment (marklubin) [inline apiana/core/components/transform/batch_inference.py]: We can just store a copy of the config dict

File: apiana/core/components/transform/batch_inference.py [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+        super().__init__(name=name or "BatchInferenceTransform", config=config_dict)
```
Comment (marklubin) [inline apiana/core/components/transform/batch_inference.py]: should be first line

File: apiana/core/components/transform/batch_inference.py [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+        for i, batch in enumerate(batches):
+            try:
+                batch_results = self._process_batch_with_retry(batch)
+                results.extend(batch_results)
+            except Exception as e:
+                error_msg = f"Failed to process batch {i+1}/{len(batches)}: {str(e)}"
+                logger.error(error_msg)
+                errors.append(error_msg)
+                # Continue with other batches or raise?
+                raise  # For now, fail fast
```
Comment (marklubin) [inline apiana/core/components/transform/batch_inference.py]: Inlcude the original exception with a message
